Added gradient line drawing to LineChartRenderer. based on PR #3142

larryonoff opened the pull request:
Some stabilization and code cleanup.

@jjatie could you please take a look at this PR. This PR stabilises and improves your PR #3142

jjatie commented:
First glance looks okay. Please fix build failures. What do you mean by "stabilization"?

I'll review thoroughly once fixed.

jjatie commented on Source/Charts/Renderers/LineChartRenderer.swift:
```diff
@@ -171,17 +191,7 @@ open class LineChartRenderer: LineRadarRenderer
-            context.setStrokeColor(drawingColor.cgColor)
-            context.strokePath()
-        }
+        drawLine(context: context, dataSet: dataSet, spline: cubicPath, matrix: valueToPixelMatrix, drawingColor: drawingColor)
```
I thought about doing this. I think this correct way to do it is
```
if dataSet.isDrawLineWithGradientEnabled
{
    drawGradientLine(context: context, dataSet: dataSet, spline: cubicPath, matrix: valueToPixelMatrix)
}
else
{
    drawLine(in: ...)
}
```

larryonoff replied:
I exposed this login into a separate function with the name `drawLine`.

jjatie replied:
I'm aware. I'm saying that `drawLine` and `drawGradientLine` should be separate methods.

larryonoff replied:
Just pushed a change. I think that this's correct now.

jjatie commented on Source/Charts/Renderers/LineChartRenderer.swift:
```diff
@@ -795,6 +806,12 @@ open class LineChartRenderer: LineRadarRenderer
     {
+        guard let gradientPositions = dataSet.gradientPositions else
+        {
+            assertionFailure("Must set `gradientPositions if `dataSet.isDrawLineWithGradientEnabled` is true")
```
👍

larryonoff commented:
@jjatie 
> First glance looks okay. Please fix build failures. What do you mean by "stabilization"?

- fixed out of bounds crash in `generateGradientLinePath` at `for i in (from + 1)..<to` by adding `guard (from + 1) < to`.
- fixed how gradient colors are picked, see `drawGradientLine`, i.e. the first data set color was always selected for points in the middle.

larryonoff commented:
I'll fix the crash. It's because different return parameters of the `func getRed(red:, green:, blue: , alpha:)` for `NSColor` and `UIColor`.

jjatie commented on Source/Charts/Renderers/LineChartRenderer.swift:
```diff
@@ -83,6 +83,19 @@ open class LineChartRenderer: LineRadarRenderer
+        context: CGContext,
+        dataSet: LineChartDataSetProtocol,
+        spline: CGMutablePath,
+        matrix: CGAffineTransform,
```
unused parameter

larryonoff replied:
fixed

larryonoff commented:
It looks that gradient locations are calculated wrong. I hope that I'll fix it tomorrow.

jjatie commented on Source/Charts/Renderers/LineChartRenderer.swift:
```diff
@@ -83,6 +83,19 @@ open class LineChartRenderer: LineRadarRenderer
+
+    private func drawLine(
+        context: CGContext,
+        dataSet: LineChartDataSetProtocol,
```
unused parameter

larryonoff replied:
fixed

larryonoff commented:
What anyone thing about implementing similar approach like for fill, i.e.
```
open class ChartGradient {
  var start: CGPoint
  var end: CGPoint
  var positions: [CGFloat]?
  var colors: [UIColor]?
}
```

jjatie commented:
That sounds like a great idea. I think the properties can all be `let` decorations. There are many areas I'd like to introduce this sort of data type.

larryonoff commented:
@jjatie what do you think about class name? is it fine? or better LineChartGradient, LineChartLineGradient?

jjatie commented:
```
@objc(ChartsGradient)
final class Gradient: NSObject
{
    let start: CGPoint
    // etc
}
```

jjatie commented on Source/Charts/Renderers/LineChartRenderer.swift:
```diff
@@ -785,6 +796,11 @@ open class LineChartRenderer: LineRadarRenderer
         // create a new path
         let to = Int(ceil(CGFloat(to - from) * phaseX + CGFloat(from)))
+
+        guard (from + 1) < to else {
```
Why do we need this?
Maybe `let from = from + 1` instead?
Would be nice to have a comment for why we use `+ 1`

larryonoff replied:
> Why do we need this?

this's because path generations starts from command `move` for the first point, and then `add line`.

jjatie commented on Source/Charts/Renderers/LineChartRenderer.swift:
```diff
@@ -795,85 +811,63 @@ open class LineChartRenderer: LineRadarRenderer
-        {
-            fatalError("Must set `gradientPositions if `dataSet.isDrawLineWithGradientEnabled` is true")
+        let boundingBox = spline.boundingBox
+        guard !boundingBox.isNull && !boundingBox.isInfinite && !boundingBox.isEmpty else {
```
comma instead of ampersand. We're listing requirements, not performing Boolean logic.

jjatie commented on Source/Charts/Utils/Platform.swift:
```diff
@@ -304,6 +320,26 @@ types are aliased to either their UI* implementation (on iOS) or their NS* imple
 
+    extension NSUIColor
+    {
+        var nsuiRGBA: (red: CGFloat, green: CGFloat, blue: CGFloat, alpha: CGFloat)? {
```
Rename to `rgba`

jjatie commented on Source/Charts/Utils/Platform.swift:
```diff
@@ -24,6 +24,22 @@ types are aliased to either their UI* implementation (on iOS) or their NS* imple
+
+    extension NSUIColor
+    {
+        var nsuiRGBA: (red: CGFloat, green: CGFloat, blue: CGFloat, alpha: CGFloat)? {
```
`rgba`

larryonoff commented:
@jjatie I think it would be great to rebase this PR into `4.0.0` branch.

petester42 commented:
Thanks 👍
